### PR TITLE
Validate regular expressions as ECMA-262

### PR DESCRIPTION
**Pull request by gjtorikian**
Here's the thing: JSON schema expects regular expression patterns to be [ECMA-262 compatible](http://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.2.3).

I wrote a little gem to detect whether a given string of RegEx is valid under ECMA-262 or not. All of the heavy lifting is done by @ammar's amazing [regexp_parser](https://github.com/ammar/regexp_parser) library. The gem is simple: given a string, does it contain any non-valid syntax? If so, it fails the validation.

The goal of this PR is to minimize false Ruby-style regular expressions from slipping through while constructing JSON schemas. Although Ruby's Regexp is being used to construct the regular expressions, it's more correct to expect them to be ECMA-262 compatible. 

I am up for any and all suggestions for improving this!

### CODE REVIEW

**Comment by brandur** [comment]
@gjtorikian Oh wow, great find on this one! Thanks. Just of out curiosity: did you actually run into a problem with a non-ECMA-262 compatible regular expression yourself? (that is to say, what spurred you to work on this feature?) It seems that parsing regular expressions using a Ruby-based third party library will introduce quite a bit of overhead compared to Ruby's built-in implementation, and I just want to make sure that the trade-off is worth it for the general case.

One alternative option that we might have is to remove the new dependency, but add support for regexp_parser and an option to turn it on. A user who needs this can then require the library themselves and enable the option.

**Comment by gjtorikian** [comment]
> did you actually run into a problem with a non-ECMA-262 compatible regular expression yourself? (that is to say, what spurred you to work on this feature?)

In the wild? As of yet, no. But we didn't want to risk the potential for accidentally defining a pattern in Ruby regex. If the schema is written in JSON, then the pattern should adhere to the JSON schema rules. Them's the breaks. 

One other thought: ideally, the schema author, not the individual submitting data, should be made aware that the schema is at fault. Perhaps moving this into `JsonSchema.parse!`, rather than `validate` would be better. I wasn't sure where in the code to place that, though. 

> One alternative option that we might have is to remove the new dependency, but add support for regexp_parser and an option to turn it on.

Yep, I'm amenable to that, too. That sounds like a better way to go, actually. Would you prefer a check like `if EcmaReValidator.defined?` ?

**Comment by brandur** [comment]
> One other thought: ideally, the schema author, not the individual submitting data, should be made aware that the schema is at fault. Perhaps moving this into JsonSchema.parse!, rather than validate would be better. I wasn't sure where in the code to place that, though. 

Good idea. If you check [this section here](https://github.com/brandur/json_schema/blob/b37922e6a221d258ae3bc00f8dc29fa8cbbfbb1e/lib/json_schema/parser.rb#L290-L310) you can see that the parser runs through each standard schema construct (the `validate_type` stuff), then moves on to perform some slightly more complicated operations for any sections that require it (the `parse_*`). We could probably add a section for `parse_pattern` with the check in it.

> Yep, I'm amenable to that, too. That sounds like a better way to go, actually. Would you prefer a check like if EcmaReValidator.defined? ?

That was my original thought too, but hmm, I guess it's somewhat non-ideal in that you get an implicit change in behavior based on an unrelated module. I can imagine a situation where someone includes the ECMA validator for something else, and then their schemas mysteriously become invalid. That said, the rate at which these two gems actually across paths in the real world probably makes this a pretty negligibly probable case. What do you think?

**Comment by gjtorikian** [comment]
> What do you think?

I've seen other projects use a more explicit option key, based on the project. I think Faraday does this pretty well, by letting you [swap adapters](https://github.com/lostisland/faraday#advanced-middleware-usage), depending on whether you want to use Ruby's `Net::HTTP`, Typhoeus, Patron, etc. Elasticsearch's [Ruby library](https://github.com/elastic/elasticsearch-ruby/blob/49bc54b2c85694968745af3469290245406b8315/elasticsearch-transport/lib/elasticsearch/transport/client.rb#L172-L183) does this pretty swell too.

That said, I'm totally cool with making this a bit more explicit. So perhaps:

* Remove the dependency on my `ecma-re-validator` gem
* Require an option to be passed in, like `:validate_with => :'ecma-re-validator'`
* If the option is passed, decide what to do:

  ``` ruby
  case opts[:validate_with]
  when :'ecma-re-validator'
   EcmaReValidator.valid?(schema.pattern)
  end
  ```

Basically I would want any validator to be used, should someone come up with something better. And if you don't care about the validation, no worries! *And* if you included the gem at some other point in your project, it wouldn't interfere.

**Comment by brandur** [comment]
@gjtorikian +1! Good call looking around for precedent on this one.

I think overall, I'm okay with either of these two approaches. I really like the Faraday-style adapter pattern, but admit that in this case, a simple `defined?` check might be the easiest option and unlikely to be too problematic in practice.

Continuing for your interface prototype for an explicit option, we might consider a global config block:

``` ruby
JsonSchema.config do |c|
  c.validate_regex_with: :'ecma-re-validator'
end
```

This should save us some time because we wouldn't have to retool the validator's interface at all, and should be "good enough" for almost all uses because it's unlikely that anyone would want to tweak this on a per-validator basis.

Anyway, I don't have a really strong opinion on this one :) Thoughts?

**Comment by gjtorikian** [comment]
All right, I've pushed up some changes to my fork demonstrating some of the results of the conversation.

**Comment by brandur** [comment]
@gjtorikian Looks good to me! Thank-you.

**Comment by gjtorikian** [comment]
Glad you liked it! Mind a gem bump? :pray:

**Comment by brandur** [comment]
@gjtorikian Yep! Pushed as 0.6.0.
